File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

dx, dy = [-1, 1, 0, 0], [0, 0, -1, 1]
def solution(maps):
    n = len(maps[0])
    m = len(maps)
    temp_map = [[0 for _ in range(n)] for _ in range(m)]
    queue = deque([(0, 0)])
    while queue:
        y, x = queue.popleft()
        if y == m - 1 and x == n - 1:
            return temp_map[y][x] + 1
        for i in range(4):
            ny, nx = y + dy[i], x + dx[i]
            if 0 <= ny < m and 0 <= nx < n and maps[ny][nx] == 1 and temp_map[ny][nx] == 0:
                temp_map[ny][nx] = temp_map[y][x] + 1
                queue.append((ny, nx))
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
using namespace std;
#define X first
#define Y second
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};

int solution(vector<vector<int> > maps)
{
    int n = maps.size();
    int m = maps[0].size();
    int dist[n][m];
    for (int i = 0; i < n; i++) {
        fill(dist[i], dist[i] + m, 0);
    }
    queue<pair<int, int>> Q;
    Q.push({0, 0});
    dist[0][0] = 1;
    while (!Q.empty()) {
       	auto cur = Q.front();
        Q.pop();
        for (int i = 0; i < 4; i++) {
            int nx = cur.X + dx[i];
            int ny = cur.Y + dy[i];
            if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
            if (dist[nx][ny] > 0 || maps[nx][ny] == 0) continue;
            dist[nx][ny] = dist[cur.X][cur.Y] + 1;
            Q.push({nx, ny});
        }
    }
    if (dist[n - 1][m - 1] > 0) return dist[n - 1][m - 1];
    else return -1;
}